File: BE/models/vendor.js
const mongoose = require('mongoose');

const vendorSchema = mongoose.Schema({
  fullName: {
    type: String,
    required: true,
    trim: true,
  },

  email: {
    type: String,
    required: true,
    trim: true,
    validate: {
      validator: (value) => {
        const result = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|.(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return result.test(value);
      },
      message: "Vui lòng nhập địa chỉ email hợp lệ",
    },
  },

  phone: {
    type: String,
    required: true,
    trim: true,
    validate: {
      validator: (value) => /^0\d{9}$/.test(value),
      message: "Số điện thoại phải có 10 chữ số và bắt đầu bằng số 0",
    },
  },


  address: {
    type: String,
    default: "",
    trim: true,
  },


  storeName:{
    type:String,
    required:true,
  },
  storeImage:{
    type:String,
    required:true,
  },
  storeDescription:{
    type:String,
    required:true,
  },


  password: {
    type: String,
    required: true,
    validate: {
      validator: (value) => value.length >= 8,
      message: "Mật khẩu phải có ít nhất 8 ký tự",
    },
  },

  role: {
    type: String,
    required: true,
    default: "vendor",
  },
});

const Vendor = mongoose.model("Vendor", vendorSchema);

module.exports = Vendor;

File: user_app/lib/data/model/vendor.dart
import 'dart:convert';

class Vendor {
  final String id;
  final String fullName;
  final String email;
  final String phone;
  final String image;
  final String address;
  final String role;
  final String password;
  final String token;
  final String? storeImage;
  final String? storeDescription;

  Vendor({
    required this.id,
    required this.fullName,
    required this.email,
    required this.phone,
    required this.image,
    required this.address,
    required this.role,
    required this.password,
    required this.token,
    this.storeImage,
    this.storeDescription,
  });

  // Chuyển đổi từ Object sang Map (Gửi API)
  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      '_id': id,
      'fullName': fullName,
      'email': email,
      'phone': phone,
      'image': image,
      'address': address,
      'role': role,
      'password': password,
      'token':token,
      'storeImage':storeImage,
      'storeDescription':storeDescription,
    };
  }

  // Chuyển đổi từ Map sang JSON (Gửi API)
  String toJson() => json.encode(toMap());

  // Chuyển đổi từ Map sang Object (Nhận API)
  factory Vendor.fromJson(Map<String, dynamic> map) {
    return Vendor(
      id: map['_id'] as String? ?? '',
      fullName: map['fullName'] as String? ?? '',
      email: map['email'] as String? ?? '',
      phone: map['phone'] as String? ?? '',
      image: map['image'] as String? ?? '',
      address: map['address'] as String? ?? '',
      role: map['role'] as String? ?? '',
      password: map['password'] as String? ?? '',
      token: map['token'] as String? ?? '',
      storeImage: map['storeImage'] as String? ?? '',
      storeDescription: map['storeDescription'] as String? ?? '',



    );
  }

  // Chuyển đổi từ JSON sang Object (Nhận API)
  factory Vendor.fromJsonString(String source) => Vendor.fromJson(json.decode(source) as Map<String, dynamic>);
}
